Userland/Libraries/LibJS/Runtime/Value.cpp

```diff
@@ -995,29 +1048,8 @@ ThrowCompletionOr<Value> add(GlobalObject& global_object, Value lhs, Value rhs)
     auto lhs_primitive = TRY(lhs.to_primitive(global_object));
     auto rhs_primitive = TRY(rhs.to_primitive(global_object));
 
-    if (lhs_primitive.is_string() && rhs_primitive.is_string()) {
-        auto const& lhs_string = lhs_primitive.as_string();
-        auto const& rhs_string = rhs_primitive.as_string();
-
-        if (lhs_string.has_utf16_string() && rhs_string.has_utf16_string()) {
-            auto const& lhs_utf16_string = lhs_string.utf16_string();
-            auto const& rhs_utf16_string = rhs_string.utf16_string();
-
-            Vector<u16, 1> combined;
-            combined.ensure_capacity(lhs_utf16_string.length_in_code_units() + rhs_utf16_string.length_in_code_units());
-            combined.extend(lhs_utf16_string.string());
-            combined.extend(rhs_utf16_string.string());
-            return Value(js_string(vm.heap(), Utf16String(move(combined))));
-        }
-    }
-    if (lhs_primitive.is_string() || rhs_primitive.is_string()) {
-        auto lhs_string = TRY(lhs_primitive.to_string(global_object));
-        auto rhs_string = TRY(rhs_primitive.to_string(global_object));
-        StringBuilder builder(lhs_string.length() + rhs_string.length());
-        builder.append(lhs_string);
-        builder.append(rhs_string);
-        return Value(js_string(vm, builder.to_string()));
-    }
+    if (lhs_primitive.is_string() || rhs_primitive.is_string())
+        return concatenate_strings(global_object, lhs_primitive, rhs_primitive);
```

Comment:
It's weird that this can throw IMO, and violates the spec's `where each argument is a String value, a code unit, or a sequence of code units` in `string-concatenation`.

Should be:

```cpp
// (In 13.15.3)
// c. If Type(lprim) is String or Type(rprim) is String, then
if (lhs_primitive.is_string() || rhs_primitive.is_string()) {
    // i. Let lstr be ? ToString(lprim).
    auto lhs_string = TRY(lhs_primitive.to_string(global_object));

    // ii. Let rstr be ? ToString(rprim).
    auto rhs_string = TRY(rhs_primitive.to_string(global_object));

    // iii. Return the string-concatenation of lstr and rstr.
    return js_string(vm, concatenate_strings(lhs_string, rhs_string));
}
```

Reply:
Hmm I was trying to keep the both-strings-are-UTF-16 case inside the `concatenate_strings` method, so that outside callers don't have to check for it themselves. We could overload `concatenate_strings` for UTF-16 strings, but then each caller still has to do something like:

```c++
if (lhs_primitive.is_string() && rhs_primitive.is_string()) {
    auto const& lhs_string = lhs_primitive.as_string();
    auto const& rhs_string = rhs_primitive.as_string();

    if (lhs_string.has_utf16_string() && rhs_string.has_utf16_string()) {
        auto const& lhs_utf16_string = lhs_string.utf16_string();
        auto const& rhs_utf16_string = rhs_string.utf16_string();
        return js_string(vm, concatenate_strings(lhs_utf16_string, rhs_utf16_string));
    }
}

if (lhs_primitive.is_string() || rhs_primitive.is_string()) {
    auto lhs_string = TRY(lhs_primitive.to_string(global_object));
    auto rhs_string = TRY(rhs_primitive.to_string(global_object));
    return js_string(vm, concatenate_strings(lhs_string, rhs_string));
}
```

(At the very least, `String.prototype.concat` will need to use this method as well. The phrase "the string-concatenation" appears 112 times in ECMA-262 so I was going to go through them all later and see what needs fixing in a separate PR)

Reply:
Something like this works:
```c++
    if (lhs_primitive.is_string() || rhs_primitive.is_string()) {
        auto lhs_string = TRY(lhs_primitive.to_primitive_string(global_object));
        auto rhs_string = TRY(rhs_primitive.to_primitive_string(global_object));
        return concatenate_strings(global_object, *lhs_string, *rhs_string);
    }
```

The downside being that we allocate a cell for non-string values (probably not a huge deal since we're already allocating a string anyways too).

Reply:
Seems like a good compromise. Thanks!